drivers/i2c/i2c_npcx_controller.c

```diff
@@ -0,0 +1,878 @@
+#define NPCX_I2C_BANK_FIFO   1
+
+/* Timeout for device should be available after reset (SMBus spec. unit:ms) */
+#define I2C_MAX_TIMEOUT 35
+
+/* Timeout for SCL held to low by slave device . (SMBus spec. unit:ms). */
+#define I2C_MIN_TIMEOUT 25
+
+/* Default maximum time we allow for an I2C transfer (unit:ms) */
+#define I2C_TRANS_TIMEOUT K_MSEC(100)
+
+/*
+ * NPCX I2C module that supports FIFO mode has 32 bytes Tx FIFO and
+ * 32 bytes Rx FIFO.
+ */
+#define NPCX_I2C_FIFO_MAX_SIZE 32
+
+/* Valid bit fields in SMBST register */
+#define NPCX_VALID_SMBST_MASK ~(BIT(NPCX_SMBST_XMIT) | BIT(NPCX_SMBST_MASTER))
+
+/*
+ * Internal SMBus Interface driver states values, which reflect events
+ * which occurred on the bus
+ */
+enum smb_oper_state {
+	NPCX_I2C_IDLE,
+	NPCX_I2C_WAIT_START,
+	NPCX_I2C_WAIT_RESTART,
+	NPCX_I2C_WRITE_FIFO,
+	NPCX_I2C_WRITE_SUSPEND,
+	NPCX_I2C_READ_FIFO,
+	NPCX_I2C_READ_SUSPEND,
+	NPCX_I2C_WAIT_STOP,
+	NPCX_I2C_ERROR_RECOVERY,
+};
+
+/* I2C timing configuration for each i2c speed */
+struct npcx_i2c_timing_cfg {
+	uint8_t HLDT; /* i2c hold-time (Unit: clocks) */
+	uint8_t k1; /* k1 = SCL low-time (Unit: clocks) */
+	uint8_t k2; /* k2 = SCL high-time (Unit: clocks) */
+};
+
+/* Device config */
+struct i2c_ctrl_config {
+	uintptr_t base; /* i2c controller base address */
+	struct npcx_clk_cfg clk_cfg; /* clock configuration */
+	int irq; /* i2c controller irq */
+	int index; /* i2c controller index */
+};
+
+/* Driver data */
+struct i2c_ctrl_data {
+	struct k_sem lock_sem; /* mutex of i2c controller */
+	struct k_sem sync_sem; /* semaphore usde for synchronization */
+	uint32_t bus_freq; /* operation freq of i2c */
+	enum smb_oper_state oper_state; /* controller operation state */
+	int trans_err;  /* error code during transaction */
+	struct i2c_msg *msg; /* current i2c msg */
+	int is_write; /* direction of current msg */
+	uint8_t *ptr_msg; /* current msg pointer for FIFO read/write */
+	uint16_t addr; /* slave address of transcation */
+	uint8_t port; /* current port used the controller */
+};
+
+/* Driver convenience defines */
+#define DRV_CONFIG(dev) \
+	((const struct i2c_ctrl_config *)(dev)->config)
+
+#define DRV_DATA(dev) \
+	((struct i2c_ctrl_data *)(dev)->data)
+
+#define HAL_I2C_INSTANCE(dev) \
+	(struct smb_reg *)(DRV_CONFIG(dev)->base)
+
+#define HAL_I2C_FIFO_INSTANCE(dev) \
+	(struct smb_fifo_reg *)(DRV_CONFIG(dev)->base)
+
+/* Recommended I2C timing values are based on 15 MHz */
+static const struct npcx_i2c_timing_cfg npcx_speed_confs[] = {
+	[NPCX_I2C_BUS_SPEED_100KHZ] = {.HLDT = 0, .k1 = 75, .k2 = 0},
+	[NPCX_I2C_BUS_SPEED_400KHZ] = {.HLDT = 7, .k1 = 24, .k2 = 18,},
+	[NPCX_I2C_BUS_SPEED_1MHZ] = {.HLDT  = 7, .k1 = 14, .k2 = 10,},
+};
+
+/* I2C controller inline functions operate in 'Normal' mode */
+static inline void i2c_ctrl_bank_select(const struct device *dev, int bank)
+{
+	struct smb_reg *const inst = HAL_I2C_INSTANCE(dev);
+
+	if (bank)
+		inst->SMBCTL3 |= BIT(NPCX_SMBCTL3_BNK_SEL);
+	else
```

Comment:
I believe Zephyr standard is braces around all single-line statements.

Reply:
Done.